### ISSUE TITLE
Code refactoring

### ISSUE THREAD

**Issue by mayacakmak** [hcrlab/stretch_web_interface#9]
Overall the code is not very modular and object-oriented. Several relevant files are very large. At some point it would be useful to refactor to change the structure a little bit to make adding new features and documentation easier. This is a bit vague at the moment so not jumping into it but will take notes in this issue of how we might be able to refactor as we continue to implement features and things become a bit more concrete.

**Comment by hello-ck** [comment]
Hi Maya!

I'm very interested in improving the quality of the code. I'd love to discuss this issue with you and your team. It would also be great to have a plan to help us coordinate refactoring. 

The aspect of the code that I'm least happy about is the lack of modularity. There are many variables and functions with scopes that are too large and unclear. This is primarily due to my lack of familiarity with JavaScript. This also used to be a single-developer, get-things-running-asap effort with the hope of someday revisiting it to improve the quality. With your lab using the web interface code so successfully and improving it in so many ways, there's real incentive to make things better!

Best wishes, 
Charlie

**Comment by mayacakmak** [comment]
Hi Charlie,

I completely agree! I actually like the approach of getting things to work without thinking too much about modularity in the beginning, because it is easy to go down the wrong path and waste a lot of time trying to make things super well structured from the beginning. Though JS is not great for modularity in general.

I'm going about it on a need basis, basically trying to make things more reusable when I realize I'm about to copy paste a large chunk of code :) For example, I just completed a major refactoring of [operator.html](https://github.com/hcrlab/stretch_web_interface/blob/master/operator/operator.html) and the coupled [operator_ui_regions.js](https://github.com/hcrlab/stretch_web_interface/blob/master/operator/operator_ui_regions.js) creating classes for `VideoControl` that has a camera video background and an interactive `Overlay` composed of `Region`s. Both files have reduced significantly in lines of code and now it is a lot easier for me to make multiple camera views with different overlays as we try to switch to the two mode (navigate-manipulate) design, as well as integrate multiple camera views.

I think there are some other opportunities throughout, but refactoring usually means breaking things before making them work again, which can sometimes take a lot of time so I'm constantly weighing whether it is worth it or not (or rather whether it is urgent or not). I'll just keep noting down ideas for refactoring in this thread and more than happy to discuss further!

**Comment by KaviMD** [comment]
Per our previous discussion about refactoring, I put together this draft plan

### Clean Up Files

- assuming the pug code is unused, remove it
- update ui_elements/operator_ui_test.html
- rename images → README/images
- add images/js/css subfolders in operator and robot
- get rid of unused commands in ros_connect.js and commands.js
    - there are a bunch of commands in `wristCommands`, and `gripperCommands` in commands that I'm not sure if we're using

### Smaller Fixes/Features

- try moving google logo into /public to fix the oath image issue
- use only 1 global `RequestResponseHandler` per interface
- switch to using `throw` for all errors
- add deadman's switch to the whole system
    - robot stop on operator disconnect
- add deadman's switch to pose saving
- fix firebase rules

### Larger Refactors

**Robot/Operator Communication**

- currently, `shared/send_recv_av.js`, `shared/request_response.js`, `operator/operator_acquire_av.js`, `robot/robot_acquire_av.js`, and `signaling_sockets.js` all contain communication code
- it might be good to move code around, so there is 3 communication files
    1. a shared file that doesn't execute any code, but contains helper functions and classes
    2. a `robot` file that sets up communication with the operator using a combination of the shared file and custom code
    3. an `operator` file that sets up communication with the robot using a combination of the shared file and custom code 

**Commands**

- fix unnecessary code duplication
    - right now there is a separate function for each possible command, it seems like there should be a cleaner way
- remove naming conflicts between ros_connect.js and commands.js
    - do both robot and operator need to import commands.js? it feels like that functionality should be split up

### Recreate the Interface using a JS framework (ex React)

Right now we are updating a ton of HTML using JS manually (the overlay regions, settings page etc.) The state of the interface in JS is completely decoupled from the HTML, which has to be manually updated. For example, had to write a bunch of code to automatically add new pose buttons when they were added to firebase, and I'm working on code to remove the buttons when they are deleted from firebase too. Using a javascript framework would automatically help keep the UI and the JS state in sync, decreasing the amount of code we have to write, hopefully also leading to less bugs, and more modularity.

**Comment by KaviMD** [comment]
The core communication refactoring is done. I'm working on documenting everything and renaming a few functions to help it make more sense

These are a few smaller bugs and other changes that we may want to address
- [ ] reach overlay doesn't enable/disable correctly
- [ ] firebase only lets you sign in if the URL starts with `localhost`, so if the operator.html is loaded on any device other than the
- [ ] firebase tries to log data even if the user is not logged in
- [ ] There are some variables in `webrtc_utils,js` that don't need to be accessed by both the robot and the operator (those should be moved into the respective `[robot, operator]_communication.js` files)
- [ ] Documentation should be added to the rest of the code in the webrtc communication files
- [ ] unused code should be commented out of `commands.js`
- [ ] all of the discrete (one time) actions in `commands.js` and `ros_connect.js` should be combined into one to decrease code duplication

**Comment by nickswalker** [comment]
With a lot of additional intermediary steps, we've arrived at a much enhanced state:
* We put everything into ES6 modules, reducing the amount of shared state
* We reformatted UI code into web components where appropriate. The division of responsibilities in the codebase roughly looks like model-view-controller now.
* We switched to Typescript (#41), which made us to annotate the code with type information. This makes some bugs harder to write, and also immediately makes it easier to see what code does based on its type signature
